File: 0853-car-fleet/0853-car-fleet.js
/**
 * @param {number} target
 * @param {number[]} position
 * @param {number[]} speed
 * @return {number}
 */
var carFleet = function (target, positions, speeds) {
  if (positions.length !== speeds.length) throw new Error("invalid positions and speeds");

  const stitchedVectors = [];

  for (let i = 0; i < positions.length; i++) {
    stitchedVectors.push([positions[i], speeds[i]]);
  }

  stitchedVectors.sort((a, b) => a[0] - b[0]);

  let fleetsCount = 0;
  let timeToTarget = -1;

  while (stitchedVectors.length !== 0) {
    const currVector = stitchedVectors.pop();

    const tempTimeToTarget = (target - currVector[0]) / currVector[1];
    if (tempTimeToTarget > timeToTarget) {
      fleetsCount++;
      timeToTarget = tempTimeToTarget;
    }
  }

  return fleetsCount;
};

File: 0853-car-fleet/0853-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: list[int], speed: list[int]) -> int:
        cars_pos_and_speed = []

        for i in range(len(position)):
            cars_pos_and_speed.append(
                (position[i], speed[i])
            )

        cars_pos_and_speed.sort(key=lambda x: -1 * x[0])
        fleet_arrival_times = []

        # print("cars_post_and_speed is", cars_pos_and_speed)

        for (curr_pos, curr_speed) in cars_pos_and_speed:
            curr_time_of_arrival = (target - curr_pos) / curr_speed
            # print("curr_time_of_arrival is", curr_time_of_arrival)
            curr_fleet_time_of_arrival = (
                fleet_arrival_times[-1]
                if fleet_arrival_times
                else float('-inf')
            )
            if curr_time_of_arrival > curr_fleet_time_of_arrival:
                fleet_arrival_times.append(curr_time_of_arrival)
                # print("appended to fleet_arrival_times")
            else:  # curr_time_of_arrival >= curr_fleet_time_of_arrival
                # print("not appended to fleet_arrival_times")
                continue

        return len(fleet_arrival_times)